spring-cloud-function-context/test/java/org/springframework/cloud/function/context/ContextFunctionCatalogAutoConfigurationTests.java

```diff
@@ -0,0 +1,159 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import reactor.core.publisher.Flux;
+
+/**
+ * @author Dave Syer
+ *
+ */
+public class ContextFunctionCatalogAutoConfigurationTests {
+
+	private ConfigurableApplicationContext context;
+	private InMemoryFunctionCatalog catalog;
+
+	@After
+	public void close() {
+		if (context != null) {
+			context.close();
+		}
+	}
+
+	@Test
+	public void simpleFunction() {
+		create(SimpleConfiguration.class);
+		assertThat(context.getBean("function")).isInstanceOf(Function.class);
+		assertThat(catalog.lookupFunction("function")).isInstanceOf(Function.class);
+	}
+
+	@Test
+	public void simpleSupplier() {
+		create(SimpleConfiguration.class);
+		assertThat(context.getBean("supplier")).isInstanceOf(Supplier.class);
+		Supplier<Flux<String>> supplier = catalog.lookupSupplier("supplier");
+		assertThat(supplier.get().blockFirst()).isEqualTo("hello");
+	}
+
+	@Test
+	public void simpleConsumer() {
+		create(SimpleConfiguration.class);
+		assertThat(context.getBean("consumer")).isInstanceOf(Consumer.class);
+		Consumer<Flux<String>> consumer = catalog.lookupConsumer("consumer");
+		consumer.accept(Flux.just("foo", "bar"));
+		assertThat(context.getBean(SimpleConfiguration.class).list).hasSize(2);
+	}
+
+	@Test
+	public void qualifiedBean() {
+		create(QualifiedConfiguration.class);
+		assertThat(context.getBean("function")).isInstanceOf(Function.class);
+		assertThat(catalog.lookupFunction("function")).isNull();
+		assertThat(catalog.lookupFunction("other")).isInstanceOf(Function.class);
+	}
+
+	@Test
+	public void aliasBean() {
+		create(AliasConfiguration.class);
+		assertThat(context.getBean("function")).isInstanceOf(Function.class);
+		assertThat(catalog.lookupFunction("function")).isNotNull();
+		assertThat(catalog.lookupFunction("other")).isInstanceOf(Function.class);
+	}
+
+	@Test
+	public void registrationBean() {
+		create(RegistrationConfiguration.class);
+		assertThat(context.getBean("function")).isInstanceOf(Function.class);
+		assertThat(catalog.lookupFunction("function")).isNull();
+		assertThat(catalog.lookupFunction("registration")).isNull();
+		assertThat(catalog.lookupFunction("other")).isInstanceOf(Function.class);
+	}
+
+	private void create(Class<?>... types) {
+		context = new SpringApplicationBuilder((Object[]) types).run();
+		catalog = context.getBean(InMemoryFunctionCatalog.class);
+	}
+
+	@EnableAutoConfiguration
+	@Configuration
+	protected static class SimpleConfiguration {
+		private List<String> list = new ArrayList<>();
+		@Bean
+		public Function<String, String> function() {
+			return value -> value.toUpperCase();
+		}
+		@Bean
+		public Supplier<String> supplier() {
+			return () -> "hello";
+		}
+		@Bean
+		public Consumer<String> consumer() {
+			return value -> list.add(value);
+		}
+	}
+
+	@EnableAutoConfiguration
+	@Configuration
+	protected static class QualifiedConfiguration {
+		@Bean
+		@Qualifier("other")
```

Comment:
Yeah, I'm not sure on the name, so I left the test as `@Qualifier`. I think it would be good to have a custom annotation. Just not convinced on the name yet.